File: 2019/05/05.clj
(ns dec02
    (:require [clojure.string :as str]))

(defn parse-file [file-name]
    (vec (map read-string (str/split (slurp file-name) #","))))

(defn parse-instruction [n]
  [(mod n 100)
   (int (/ (mod n 1000) 100))
   (int (/ (mod n 10000) 1000))
   (int (/ (mod n 100000) 10000))])

(defn positive-position [_ position] (pos? position))

(defn read-from [program position mode]
  (case mode
    0 (nth program (nth program position))
    1 (nth program position)))

(defn write-to [program position value]
  (let [pos (nth program position)]
    (assoc program pos value)))

(defn handle-position [program position input output]
    (let [[op-code mode-a mode-b mode-c] (parse-instruction (nth program position))]
        (let [[program' position' output'] (case op-code
          99 [program -1 output]
          1 [(write-to program (+ position 3)
                (+
                  (read-from program (+ position 1) mode-a)
                  (read-from program (+ position 2) mode-b)))
              (+ position 4)
              output]
          2 [(write-to program (+ position 3)
                (*
                  (read-from program (+ position 1) mode-a)
                  (read-from program (+ position 2) mode-b)))
              (+ position 4)
              output]
          3 [(write-to program (+ position 1) input)
             (+ position 2)
             output]
          4 [program
             (+ position 2)
             (read-from program (+ position 1) mode-a)]
          5 [program
             (if (= (read-from program (+ position 1) mode-a) 0)
               (+ position 3)
               (read-from program (+ position 2) mode-b))
             output]
          6 [program
             (if (= (read-from program (+ position 1) mode-a) 0)
               (read-from program (+ position 2) mode-b)
               (+ position 3))
             output]
          7 [(write-to program (+ position 3)
                (if (<
                     (read-from program (+ position 1) mode-a)
                     (read-from program (+ position 2) mode-b))
                  1 0))
             (+ position 4)
             output]
          8 [(write-to program (+ position 3)
                (if (=
                     (read-from program (+ position 1) mode-a)
                     (read-from program (+ position 2) mode-b))
                  1 0))
             (+ position 4)
             output])]

      [program' position' output'])))

(defn run-program [program input]
      (loop [program program position 0 output 0]
          (let [[program' position' output'] (handle-position program position input output)]
            (if (neg? position')
              output'
              (recur program' position' output')))))

(let [program (parse-file "program.testcode")]        
    (println "Part 1:" (run-program program 1)))

(let [program (parse-file "program.testcode")]        
    (println "Part 2:" (run-program program 5)))

File: 2019/05/05.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func parseFile(fileName string) []int {
	data, _ := ioutil.ReadFile(fileName)
	program := []int{}
	for _, s := range strings.Split(string(data), ",") {
		trimmed := strings.Trim(s, "\n")
		if i, err := strconv.Atoi(trimmed); err == nil {
			program = append(program, i)
		}
	}
	return program
}

type Memory struct {
	data []int
}

type Mode int

const (
	Immediate = 1
	Position  = 0
)

type OpCode int

const (
	Add         = 1
	Multiply    = 2
	Input       = 3
	Output      = 4
	JumpIfTrue  = 5
	JumpIfFalse = 6
	LessThan    = 7
	Equals      = 8
	Halt        = 99
)

func (m *Memory) read(position int, mode Mode) int {
	switch mode {
	case Immediate:
		return m.data[position]
	case Position:
		return m.data[m.data[position]]
	}
	return 0
}

func (m *Memory) write(position int, value int) {
	m.data[m.data[position]] = value
}

func parseInstruction(instruction int) (OpCode, Mode, Mode) {
	opCode := OpCode(instruction % 100)
	modeA := Mode((instruction % 1000) / 100)
	modeB := Mode((instruction % 10000) / 1000)
	return opCode, modeA, modeB
}

func runProgram(memory Memory, input int) int {
	position := 0
	output := 0

	for {
		instruction := memory.read(position, Immediate)
		opCode, modeA, modeB := parseInstruction(instruction)

		switch opCode {
		case Halt:
			return output
		case Add:
			a := memory.read(position+1, modeA)
			b := memory.read(position+2, modeB)
			memory.write(position+3, a+b)
			position += 4
		case Multiply:
			a := memory.read(position+1, modeA)
			b := memory.read(position+2, modeB)
			memory.write(position+3, a*b)
			position += 4
		case Input:
			memory.write(position+1, input)
			position += 2
		case Output:
			output = memory.read(position+1, modeA)
			position += 2
		case JumpIfTrue:
			a := memory.read(position+1, modeA)
			if a != 0 {
				position = memory.read(position+2, modeB)
			} else {
				position += 3
			}
		case JumpIfFalse:
			a := memory.read(position+1, modeA)
			if a == 0 {
				position = memory.read(position+2, modeB)
			} else {
				position += 3
			}
		case LessThan:
			a := memory.read(position+1, modeA)
			b := memory.read(position+2, modeB)
			if a < b {
				memory.write(position+3, 1)
			} else {
				memory.write(position+3, 0)
			}
			position += 4
		case Equals:
			a := memory.read(position+1, modeA)
			b := memory.read(position+2, modeB)
			if a == b {
				memory.write(position+3, 1)
			} else {
				memory.write(position+3, 0)
			}
			position += 4
		}
	}
}

func main() {
	program := parseFile("program.testcode")
	result := runProgram(Memory{data: program}, 1)
	fmt.Println("Part 1:", result)

	program2 := parseFile("program.testcode")
	result2 := runProgram(Memory{data: program2}, 5)
	fmt.Println("Part 2:", result2)
}
